File: gesture.py
"""
gesture.py

Utility functions for classifying hand gestures from MediaPipe Hands
landmarks. No ML here on purpose -- just geometric rules on the 21
landmark points, which is fast, robust, and easy to tune.

MediaPipe hand landmark indices (for reference):
    0: WRIST
    1-4:   THUMB (CMC, MCP, IP, TIP)
    5-8:   INDEX (MCP, PIP, DIP, TIP)
    9-12:  MIDDLE (MCP, PIP, DIP, TIP)
    13-16: RING (MCP, PIP, DIP, TIP)
    17-20: PINKY (MCP, PIP, DIP, TIP)
"""

import math


WRIST = 0
THUMB_CMC, THUMB_MCP, THUMB_IP, THUMB_TIP = 1, 2, 3, 4
INDEX_MCP, INDEX_PIP, INDEX_DIP, INDEX_TIP = 5, 6, 7, 8
MIDDLE_MCP, MIDDLE_PIP, MIDDLE_DIP, MIDDLE_TIP = 9, 10, 11, 12
RING_MCP, RING_PIP, RING_DIP, RING_TIP = 13, 14, 15, 16
PINKY_MCP, PINKY_PIP, PINKY_DIP, PINKY_TIP = 17, 18, 19, 20

FINGER_TIPS = [INDEX_TIP, MIDDLE_TIP, RING_TIP, PINKY_TIP]
FINGER_PIPS = [INDEX_PIP, MIDDLE_PIP, RING_PIP, PINKY_PIP]


def _dist(a, b):
    return math.hypot(a.x - b.x, a.y - b.y)


def fingers_up(landmarks, handedness_label="Right"):
    """
    Returns a list of 5 booleans [thumb, index, middle, ring, pinky]
    indicating whether each finger is extended.

    landmarks: the .landmark list from a MediaPipe HandLandmark result
               (normalized image coords, y grows downward).
    handedness_label: "Left" or "Right" as reported by MediaPipe
                       (note: MediaPipe mirrors this relative to the
                       *camera's* view, so a hand the user sees as
                       their right hand is often labeled "Left" if
                       the feed isn't flipped -- flip your frame with
                       cv2.flip(frame, 1) upstream and this lines up
                       with intuition).
    """
    result = []

    if handedness_label == "Right":
        result.append(landmarks[THUMB_TIP].x < landmarks[THUMB_IP].x)
    else:
        result.append(landmarks[THUMB_TIP].x > landmarks[THUMB_IP].x)

    for tip, pip in zip(FINGER_TIPS, FINGER_PIPS):
        result.append(landmarks[tip].y < landmarks[pip].y - 0.02)

    return result  


def _thumb_index_pinch(landmarks, pinch_threshold=0.055):
    """True if thumb tip and index tip are close together (a 'pinch'/OK sign)."""
    return _dist(landmarks[THUMB_TIP], landmarks[INDEX_TIP]) < pinch_threshold


def classify_gesture(landmarks, handedness_label="Right"):
    """
    Classifies a single hand's pose into one of a small set of named
    gestures based on which fingers are extended.

    Returns one of: "draw", "peace", "ok_sign", "three_up", "open_palm",
    "thumbs_up", "fist", "unknown"
    """
    thumb, index, middle, ring, pinky = fingers_up(landmarks, handedness_label)

  
    if middle and ring and pinky and _thumb_index_pinch(landmarks):
        return "ok_sign"  

    if index and not middle and not ring and not pinky:
        return "draw"  

    if index and middle and not ring and not pinky:
        return "peace"  

    if index and middle and ring and not pinky:
        return "three_up"  

    if index and middle and ring and pinky:
        return "open_palm"  

    if thumb and not index and not middle and not ring and not pinky:
        return "thumbs_up"  

    if not thumb and not index and not middle and not ring and not pinky:
        return "fist"  

    return "unknown"


def fingertip_pos(landmarks, frame_w, frame_h, tip=INDEX_TIP):
    """Pixel-space (x, y) position of a given fingertip landmark."""
    lm = landmarks[tip]
    return int(lm.x * frame_w), int(lm.y * frame_h)


def hand_center(landmarks, frame_w, frame_h):
    """Pixel-space centroid of the whole hand (average of all landmarks)."""
    xs = [lm.x for lm in landmarks]
    ys = [lm.y for lm in landmarks]
    cx = sum(xs) / len(xs) * frame_w
    cy = sum(ys) / len(ys) * frame_h
    return int(cx), int(cy)

File: gesture.js
/**
 * gesture.js
 *
 * Classifies hand poses from MediaPipe Hands landmarks into named
 * gestures, mirroring the logic in the Python gesture_utils.py.
 *
 * IMPORTANT: this assumes the image fed into MediaPipe has already
 * been mirrored (see draw.js, which draws the camera feed onto a
 * flipped offscreen canvas before calling hands.send). With a
 * mirrored input, MediaPipe's handedness label directly matches the
 * hand the user intuitively expects ("Right" = their right hand),
 * exactly like Python's `cv2.flip(frame, 1)` before processing.
 * If you ever feed MediaPipe an unflipped frame, you must swap the
 * handedness label AND flip the thumb-direction test below, or the
 * thumb-based gestures (thumbs_up) will only work for one hand.
 *
 * MediaPipe hand landmark indices (for reference):
 *   0: WRIST
 *   1-4:   THUMB (CMC, MCP, IP, TIP)
 *   5-8:   INDEX (MCP, PIP, DIP, TIP)
 *   9-12:  MIDDLE (MCP, PIP, DIP, TIP)
 *   13-16: RING (MCP, PIP, DIP, TIP)
 *   17-20: PINKY (MCP, PIP, DIP, TIP)
 */

const THUMB_TIP = 4, THUMB_IP = 3;
const INDEX_TIP = 8, INDEX_PIP = 6;
const MIDDLE_TIP = 12, MIDDLE_PIP = 10;
const RING_TIP = 16, RING_PIP = 14;
const PINKY_TIP = 20, PINKY_PIP = 18;

function dist(a, b) {
  return Math.hypot(a.x - b.x, a.y - b.y);
}

/**
 * Returns { thumb, index, middle, ring, pinky } booleans for whether
 * each finger is extended.
 *
 * handedness: "Left" or "Right", as reported by MediaPipe on a
 * mirrored input frame (see module docstring above).
 */
function fingersUp(lm, handedness = "Right") {
  const thumb =
    handedness === "Right"
      ? lm[THUMB_TIP].x < lm[THUMB_IP].x
      : lm[THUMB_TIP].x > lm[THUMB_IP].x;

  const index = lm[INDEX_TIP].y < lm[INDEX_PIP].y - 0.02;
  const middle = lm[MIDDLE_TIP].y < lm[MIDDLE_PIP].y - 0.02;
  const ring = lm[RING_TIP].y < lm[RING_PIP].y - 0.02;
  const pinky = lm[PINKY_TIP].y < lm[PINKY_PIP].y - 0.02;

  return { thumb, index, middle, ring, pinky };
}

function thumbIndexPinch(lm, threshold = 0.055) {
  return dist(lm[THUMB_TIP], lm[INDEX_TIP]) < threshold;
}

/**
 * Classifies a single hand's pose into one of a small set of named
 * gestures. Returns one of:
 *   "draw", "peace", "three_up", "open_palm", "ok_sign",
 *   "thumbs_up", "rock_on", "fist", "unknown"
 */
function classifyGesture(lm, handedness = "Right") {
  const { thumb, index, middle, ring, pinky } = fingersUp(lm, handedness);

  // OK sign checked first: thumb+index pinched, other three up.
  if (middle && ring && pinky && thumbIndexPinch(lm)) return "ok_sign";

  if (index && !middle && !ring && !pinky) return "draw";
  if (index && middle && !ring && !pinky) return "peace";
  if (index && middle && ring && !pinky) return "three_up";
  if (index && middle && ring && pinky) return "open_palm";
  if (thumb && !index && !middle && !ring && !pinky) return "thumbs_up";
  if (index && pinky && !middle && !ring) return "rock_on";
  if (!thumb && !index && !middle && !ring && !pinky) return "fist";

  return "unknown";
}


function landmarkPos(lm, frameW, frameH, index = INDEX_TIP) {
  return { x: lm[index].x * frameW, y: lm[index].y * frameH };
}
